Add design doc for error handling (#804)
unicode-org/message-format-wg: exploration/error-handling.md

# Error Handling

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@echeran</dd>
		<dt>First proposed</dt>
		<dd>2024-06-02</dd>
        <dt>Issues</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/issues/782">#782</a></dd>
		<dt>Pull Requests</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/795">#795</a></dd>
	</dl>
</details>

## Objective

Decide whether and what implementations "MUST" / "SHOULD" / "MAY" perform after a runtime error, regarding:

1. information about error(s)
    - including, if relevant, the minimum number of errors for which such information is expected
1. a fallback representation of the message

## Background

In practice,
runtime errors happen when formatting messages.

> It is useful to provide information about any formatting error(s) back to the callsite.

Reviewer: Not just formatting errors.

```suggestion
It is useful to provide information about any errors back to the callsite.
```

> It is useful to the end user to provide best effort fallback representation of the message.

Reviewer 2: ```suggestion
It is useful to the end user to provide a best effort fallback representation of the message.
```

However, implementations of MessageFormat 2.0 will be faced with different constraints due to various reasons:

* Programming language: the language of the implementation informs idiomatic patterns of error handling.
In Java, errors are thrown and subsequently caught in `try...catch` block.
In Rust, fallible callsites (those which can return errors) should return a `Result<T, Err>` monad. 
In both languages, built-in error handling assumes a singular error.
* Environment constriants: as mentioned in [feedback from ICU4X](https://github.com/unicode-org/message-format-wg/issues/782#issuecomment-2103177417),
ICU4X operates in low resource environments for which returning at most 1 error is desirable
because returning more than 1 error would require heap allocation.
* Programming conventions and idioms: in [feedback from ICU-TC](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99),
they found over the 25 years of maintaining the library that there was more cost than benefit in providing a default best effort return value at the same time as providing error information.

> The additional constraint in ICU4C's C++ style to use error code rather than throwing errors using the STL further complicates the usefulness and likelihood to be used correctly during nested calls.

Reviewer: typo

```suggestion
The additional constraint in ICU4C's C++ style to return an error code rather than throwing errors using the STL further complicates the usefulness and likelihood to be used correctly during nested calls.
```

## Proposed Design

The following spec text is proposed:

> > In all cases, when encountering an error during formatting,
> > a message formatter MUST provide some representation of the message,
> > or MUST provide an informative error or errors.
> > An implementation MAY provide both.

Reviewer: To me, the fact of an error is more important than the fallback representation. Showing the fallback representation to end users is less good than having good error recovery (or no error at all). This wording makes the fallback representation first and thus, seemingly, the more important thing.

I also find the "MUST or MUST" formulation to be weird. I think this is trying to say "MUST provide  some representation of the message or an informative error or errors or both"

I think it would be weird if an implementation _never failed_, that is, if it returned the fallback representation silently with no signal of failure. When would this be a good thing? How would the caller find out about the error.

Note that we do not specify how errors are signaled. Just because (for example) Java often throws `Throwable` does not mean that the MF2 implementation has to use that as the error mechanism. See, for example, the use of `ParsePosition` in `NumberFormat`.

Thus I'd suggest an additional alternative:

```suggestion
> In all cases, when encountering an error,
> a message formatter MUST provide an informative error or errors.
> It MAY also provide the appropriate fallback representation of the _message_ defined
> in this specification.
```

This solution requires implementations to return _something_,
but it leaves the decision to the implementation whether to:

> * return an error (or errors)

Reviewer: I would probably say:

```suggestion
* return or emit an error or errors
```

One reason is that throwing an exception (in languages that do that) is not the same as returning a return value.

Author: I used the word "signal" instead of "emit" because that better accomodates all of the alternatives in the solution space.

Note: I had to rewrite this section to reflect the extra constraint that I am taking on from your suggestion in the other [conversation thread](https://github.com/unicode-org/message-format-wg/pull/804#discussion_r1624685199), since this section existed to clarify the spec text concretely and in plain words.

1. a fallback representation of the message

## Background

In practice,
runtime errors happen when formatting messages.
It is useful to provide information about any errors back to the callsite.
It is useful to the end user to provide a best effort fallback representation of the message.
Specifying the behavior in such cases promotes consistent results across conformant implementations.

However, implementations of MessageFormat 2.0 will be faced with different constraints due to various reasons:

* Programming language: the language of the implementation informs idiomatic patterns of error handling.
In Java, errors are thrown and subsequently caught in `try...catch` block.
In Rust, fallible callsites (those which can return errors) should return a `Result<T, Err>` monad. 
In both languages, built-in error handling assumes a singular error.
* Environment constriants: as mentioned in [feedback from ICU4X](https://github.com/unicode-org/message-format-wg/issues/782#issuecomment-2103177417),
ICU4X operates in low resource environments for which returning at most 1 error is desirable
because returning more than 1 error would require heap allocation.

> * Programming conventions and idioms: in [feedback from ICU-TC](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99),
> they found over the 25 years of maintaining the library that there was more cost than benefit in providing a default best effort return value at the same time as providing error information.

Reviewer 3: Do I understand right that this is referring to the following comments from the linked document?
> Rich: What Tim was describing was what we do in some places in ICU where we return a warning. Like our Missing Resource Warning.
> Markus: Right, and we’ve regretted that decision for years. We cooked up the warnings as a side channel so that we didn’t have to create APIs for various things. Ex: we went to a fallback language or default language instead of the request language. This is clunky because it doesn't survive nested functions or chained functions. It’s also confusing b/c we happily continue when there’s a warning code. You can have a function set a warning code, then you have a chain of 3-4 functions afterwards that are executed before we have code that looks at the error code, and then you don’t even know what the warning pertained to once you see it. And we keep Java & C++ in sync, but there’s no way to implement warnings in Java. Why do we have warnings? It’s so that the caller can decide to perform alternate behavior. However, we’ve been much more successful just giving a best effort return value that’s not exactly what the caller requested.

If that's the case, then isn't the feedback more appropriately represented as the inverse of what's summarised here?

```suggestion
* Programming conventions and idioms: in [feedback from ICU-TC](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99),
they found over the 25 years of maintaining the library that there was more cost than benefit in providing error information at the same time as providing a default best effort return value.
```

* an API that throws or returns an Error object when encountering an error
* an API that includes a read/write `ErrorCode` argument 
* two APIs, a permissive one that always returns a best-effort formatted result
and a stricter one that throws or returns an Error object when encountering an error
* an API that always returns a best-effort formatted result
and a global `boolean` values

This does not give implementations full freedom to return _nothing_ or some other behavior.

## Alternatives Considered

### Current spec: require information from error(s) and a representative best effort message

The current spec text says:

> In all cases, when encountering a runtime error,
> a message formatter MUST provide some representation of the message.
> An informative error or errors MUST also be separately provided.

This alternative places constraints on implementations to provide multiple avenues of useful information (to the callsite and user).

This alternative establishes constraints that would contravene the constraints that exist in projects that have implemented MF 2.0 (or likely will soon), based on:
* programming language idioms/constraints

> * execution environment constraints

Reviewer 3: What execution environment constraints does this alternative contravene? The only such mentioned in this document is that the cost of returning more than one error may be prohibitive in some cases, and the current text explicitly says "error or errors" to allow for an implementation signaling a single error to be valid.
```suggestion
```

Author: The text says "an informative error". As written, that implies to me an error object, like a `java.lang.Throwable` instance. Despite discussion in last week's meeting to interpret that phrase as equivalent to "whether or not an error occurred", it comes across distinctly as something more and thus should be rewritten if the intent is not so. There are applications like ICU and potentially browsers that might only want to provide a best effort message and signal an error, but not pay the cost of creating "informative error" objects each time.

As I mentioned in the 2024-04-09 meeting, another paradigm from which to look at this, besides "whether is returning an error possible", is "how actionable is returning the error object". ICU & browsers need to be performant and might not want to pay the cost of the creating a full error object.

> * experience-based programming guidelines

Reviewer 3: It is not clear to me what these guidelines are. Could they be explicitly spelled out?

Author: Done. Added text and a link to the same document as mentioned above.

they found over the 25 years of maintaining the library that there was more cost than benefit in additionally providing error information with a default best effort return value compared to just returning the default best effort value.
The additional constraint in ICU4C's C++ style to return an error code rather than throwing errors using the STL further complicates the usefulness and likelihood to be used correctly by developers, especially during nested calls.

> [!NOTE]
> The wording in this document uses the word "signal" in regards to providing
> information about an error rather than "return" or "emit" when referring to
> a requirement that an implementation must at least indicate that an error has
> occurred.
> The word "signal" better accomodates more alternatives in the solution space
> like those that only choose to indicate that an error occurred,
> while still including those that additionally prefer to return the error
> itself as an error object.
> (By contrast, "return an error" implies that an error object will be thrown or
> returned, and "emit an error" is ambiguous as to what is or isn't performed.)

> 

Reviewer: ```suggestion
## Use Cases

As a software developer, I want calls to message format to signal runtime errors
in a manner consistent with my programming language/environment.
I would like error signals to include diagnostic information that allows me to debug errors.

As a software developer, I sometimes need to be able to emit a formatted message
even if a runtime error has occurred.

As a software developer, I sometimes want to avoid "fatal" error signals,
such as might occur due to unconstrained inputs,
due to errors in translation of the message,
or other reasons outside the developer's control.
For example, in Java, throwing an Exception is a common means of signaling an error.
However, `java.text.NumberFormat` provide both throwing and non-throwing
`parse` methods to allow developers to avoid a "fatal" throw of `ParseException`
(if the exception were uncaught).

As a MessageFormat implementer, I want to be able to signal errors in an idiomatic way
for my language and still be conformant with MF2 requirements.

```

## Proposed Design

The following spec text is proposed:

> In all cases, when encountering an error,
> a message formatter MUST be able to signal an error or errors.
> It MAY also provide the appropriate fallback representation of the _message_ defined
> in this specification.

This solution requires implementations to be able to signal an error occurred,
which can be accomplished in different ways. Ex:

* an API that throws or returns an Error object when encountering an error
* an API that includes a read/write `ErrorCode` argument 
* two APIs, a permissive one that always returns a best-effort formatted result
and a stricter one that throws or returns an Error object when encountering an error
* an API that always returns a best-effort formatted result
and a global `boolean` values

This does not give implementations full freedom to return _nothing_ or some other behavior.

> ## Alternatives Considered

Reviewer 3: As discussed in https://github.com/unicode-org/message-format-wg/pull/804#issuecomment-2158131626:

```suggestion
## Alternatives Considered

### Separate Requirements

The following spec text is proposed:

> When formatting a message with one or more errors:
> - An implementation MUST provide a way for a user to be informed
>   of the name of at least one of the errors,
>   either directly or via an identifying error code.
> - An implementation MUST provide a way for a message with one or more
>   _Resolution Errors_ or _Message Function Errors_ to be formatted
>   using a fallback representation.
>
> The two above requirements MAY be fulfilled by a single formatting method,
> or separately by more than one such method.

This alternative ensures that even messages with runtime errors
can be formatted with the same string representation in all implementations.

Compared to the current spec text,
the under-specified "informative error" is replaced
with a more concrete minimum requirement,
which can be fulfilled with a two-byte error code in very limited environments.

The phrase "In all cases" is left out,
and the ability to fulfill the requirements separately
rather than at the same time is explicitly called out.
```

# Error Handling

> Status: **Proposed**

Reviewer 3: I think this is good as is, but landing this as "proposed" would be fine as well.

```suggestion
Status: **Accepted**
```

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@echeran</dd>
		<dt>First proposed</dt>
		<dd>2024-06-02</dd>
        <dt>Issues</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/issues/782">#782</a></dd>
		<dt>Pull Requests</dt>

> 		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/795">#795</a></dd>

Reviewer 3: ```suggestion
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/795">#795</a></dd>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/804">#804</a></dd>
```

they found over the 25 years of maintaining the library that there was more cost than benefit in additionally providing error information with a default best effort return value compared to just returning the default best effort value.
The additional constraint in ICU4C's C++ style to return an error code rather than throwing errors using the STL further complicates the usefulness and likelihood to be used correctly by developers, especially during nested calls.

> [!NOTE]
> The wording in this document uses the word "signal" in regards to providing
> information about an error rather than "return" or "emit" when referring to
> a requirement that an implementation must at least indicate that an error has
> occurred.
> The word "signal" better accomodates more alternatives in the solution space
> like those that only choose to indicate that an error occurred,
> while still including those that additionally prefer to return the error
> itself as an error object.
> (By contrast, "return an error" implies that an error object will be thrown or
> returned, and "emit an error" is ambiguous as to what is or isn't performed.)
## Use Cases

> As a software developer, I want calls to message format to signal runtime errors

Reviewer 4: ```suggestion
As a software developer, I want message formatting calls to signal runtime errors
```

in a manner consistent with my programming language/environment.
I would like error signals to include diagnostic information that allows me to debug errors.

As a software developer, I sometimes need to be able to emit a formatted message
even if a runtime error has occurred.

As a software developer, I sometimes want to avoid "fatal" error signals,

> such as might occur due to unconstrained inputs,
> due to errors in translation of the message,
> or other reasons outside the developer's control.

Reviewer 4: ```suggestion
such as might occur due to unconstrained inputs,
errors in translation of the message,
or other reasons outside my control.
```

However, `java.text.NumberFormat` provide both throwing and non-throwing
`parse` methods to allow developers to avoid a "fatal" throw of `ParseException`
(if the exception were uncaught).

As a MessageFormat implementer, I want to be able to signal errors in an idiomatic way
for my language and still be conformant with MF2 requirements.

## Proposed Design

The following spec text is proposed:

> A message formatter MUST signal errors required by this specification.
> It SHOULD provide the specific information for each error reported.
> A message formatter SHOULD provide a fallback representation of the message.
>
> [!NOTE]
> The fallback representation of a message MAY be provided by a separate API.

This solution requires implementations to be able to signal an error occurred,
which can be accomplished in different ways. Ex:

* an API that throws or returns an Error object when encountering an error
* an API that includes a read/write `ErrorCode` argument 
* two APIs, a permissive one that always returns a best-effort formatted result
and a stricter one that throws or returns an Error object when encountering an error
* an API that always returns a best-effort formatted result

> and a global `boolean` values

Reviewer 4: ```suggestion
and sets a global `boolean` value
```

* execution environment constraints
* experience-based programming guidelines

For example, in ICU, 
[the suggested practice](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99)

> is to avoid additionally returning optional error codes when providing best-effort formatted results.

Reviewer 3: This applies to all solutions that require a runtime error to be signalled in some way, right?

Reviewer 3: By "solution" I meant the choice of spec language that we are debating here.

As I understand it (and please do correct me if I'm wrong), the ICU experience-based programming guideline in this case is roughly speaking, "You should not report formatting errors if you can fallback to some formatted result."

My point here is that this guideline applies to all spec text choices which require an error to be reported; for example, the current proposed design states:

> In all cases, when encountering an error,
> a message formatter MUST be able to signal an error or errors.

Therefore, that choice would also go against the above-mentioned experience-based programming guidelines. Is this not the case?

Just to be clear, I disagree about following this guideline for MF2, but here I'm trying to make sure that it's logically and consistently applied as it's being used as an argument in choosing between alternative designs.

[the suggested practice](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99)
is to avoid additionally returning optional error codes when providing best-effort formatted results.

### Require a best-effort message value and signaling of an error

> In all cases, when encountering an error,
> a message formatter MUST be able to signal an error or errors.
> It MUST also provide the appropriate fallback representation of the _message_ defined
> in this specification.

> This alternative maintains that an implementation must both return a best-effort message
> and must signal that an error has occurred.
> Compared to the current text,
> this alternative slightly loosens the constraint of 
> returning an error into only needing to signal an error.

Reviewer: I think this is slightly misleading, in that it would be conformant to have two `format` methods, one that returns an error and one that returns the fallback message (cf. the conversation with @eemeli) 

Perhaps:

```suggestion
This alternative requires that an implementation provide both an error signal
and a means of accessing a "best-effort" fallback message.
This slightly relaxes the requirement of "returning" an error
(to allow a locally-appropriate signal of the error).

Under this alternative, implementations can be conformant by providing
two separate formatting methods or functions,
one of which returns the fallback string and one of which signals the error.
```
